### DISCUSSION TITLE
Starting activities from Jetpack Compose Composables with Decompose

### DISCUSSION THREAD

**Discussion by moffpage** [badoo/Decompose#130] [Q&A] [answered]
Hello! Thanks for your amazing work! I've got a question for you: How would you start activity with an Intent (e.g. to open a map or dialer) in KMM project using Decompose with Jetpack Compose for Android (at least for now 👀). Would you just start activity from @Composable function by passing LocalContext.current or delegate it to a Store from the Component? How would you retrieve the result back?
I've been using a library for routing called Modo by terrakok: https://github.com/terrakok/Modo, his library is also cross-platform, but has a different implementation than yours and the sample has a feature where he navigates to a browser web view. When setting up, you have to link ModoRender to an activity so that when you could tell router to launch ExternalScreen-s outside of the view layer. How should we handle these use cases (intent launch + getting result back) with Decompose?
Maybe an update for **extensions-compose-jetpack** ? 👀

**Comment by Shabinder** [comment]
As for now, I just pass the intent as a callback _(which can be called later)_ when inflating your component in Android's Activity.
Similar to how we pass dependencies.

I wonder if there is a better approach for this.

**Reply by moffpage** [reply]
Well, I believe you don't inflate your components in activity, that's just for the root one. You think I should pass the callback down for like even 2 subcomponents? Like e.g. if I want to start dialer from component *X*, which is a child of component *Y*, which in itself is a child of component *Z*, which is a child of the root component. Or should I start an activity while consuming events in view by sending intent to component?

**Reply by Shabinder** [reply]
>Well, I believe you don't inflate your components in activity, that's just for the root one.

>**when inflating your component**

Here I meant Root Component Only, 

As you described the relations: **_Root->Z->Y->X_**, for instance, if there is a specific event which I Need to be available in **every component**, then maybe I can consider it making available statically, and will initialize It as soon as I initializes root(use late init and pass it to root and initialize it in init{} block).
However when you want to access it in 1-2 components only then maybe you can pass the callback around.

@arkivanov do you think there is a better approach to this?

**Comment by arkivanov** [comment, marked as answer]
Personally I like explicity and compile-time safety. So I would go with the following solution:

```kotlin
// In module X

class ComponentX(
    componentContext: ComponentContext,
    private val startCamera: (CameraParams) -> Unit
) : ComponentContext by componentContext {

    fun onCameraResult(fileName: String) {
    }
}

class CameraParams
```

```kotlin
// In module Y

class ComponentY(
    componentContext: ComponentContext,
    private val startCamera: (CameraParams) -> Unit,
    private val startBrowser: (url: String) -> Unit
) : ComponentContext by componentContext {

    fun onCameraResult(fileName: String) {
        // Deliver the result to ComponentX
    }
}

class CameraParams
```

```kotlin
// In module Z

class ComponentZ(
    componentContext: ComponentContext,
    private val startCamera: (CameraParams) -> Unit,
    private val startBrowser: (url: String) -> Unit,
    private val startFileChooser: () -> Unit
) : ComponentContext by componentContext {

    fun onCameraResult(fileName: String) {
        // Deliver the result to ComponentY
    }

    fun onFileChooserResult(fileName: String) {
    }
}

class CameraParams
```

```kotlin
// In module Root

class RootComponent(
    componentContext: ComponentContext,
    private val startCamera: (CameraParams) -> Unit,
    private val startBrowser: (url: String) -> Unit,
    private val startFileChooser: () -> Unit
) : ComponentContext by componentContext {

    fun onCameraResult(fileName: String) {
        // Deliver the result to ComponentZ
    }

    fun onFileChooserResult(fileName: String) {
        // Deliver the result to ComponentZ
    }
}

class CameraParams
```

Such approach clearly indicates what screens can be launched by the whole tree or a sub-tree. You must satisfy all dependencies, otherwise it won't compile. This is specially important when you want to reuse a component, with all its children. E.g. when you include `ComponentY` in another component, it is clear that it can launch camera and browser and what are the parameters. You may choose to split screen arguments (e.g. `(a, b, c) -> Unit`, etc.), or combine them into one single class (e.g. `(CameraParams) -> Unit`). The later will require to add similar classes in each parent level and provide mappings between them. Yes, it is verbose, but at the same time it is explicit.

Later, if for example you will have `CameraComponent`, you may change your `RootComponent`:

```kotlin
// In module Root

class RootComponent(
    componentContext: ComponentContext,
    private val startBrowser: (url: String) -> Unit,
    private val startFileChooser: () -> Unit
) : ComponentContext by componentContext {

    fun onFileChooserResult(fileName: String) {
        // Deliver the result to ComponentZ
    }

    // Handle camera requests from the ComponentZ using CameraComponent
}
```

And so it will become single activity in this case.

**Reply by moffpage** [reply]
Wouldn't the root component be bloated and cluttered in such case? I don't think I quite like this solution, I'm not sure what's better otherwise.

**Reply by arkivanov** [reply]
Let's step back for a minute and think about navigation between components only. Suppose you have the following tree of components currently active:

```
  A
  |
  B
 / \
C   D
```

And let's say you also have the following tree of reusable components in your project:

```
  X
 / \
Y   Z
```

Now you want to display X (with all children) when a button is clicked in component `C`. The scope where you want `X` to be displayed will determine which level's `Router` you should use. Some possible ways are:

```
      A          A        A
      |          |        |
      B          B        X
     / \        / \      / \
    C   D      X   D    Y   Z
   /          / \
  X          Y   Z
 / \
Y   Z
```

With the first option it is straightforward, you just use `C`'s own `Router` and handle the navigation internally.

But with other options `C` can't handle the navigation internally. Another constraint here is that `C` should not make any assumptions about its parents, it should not accept parent's `Router`, or anything like this. The navigation on this button click is beyond `C`'s responsibility. So `C` should just signal to its parent an event (or callback), like `onFinished`, `onPhotoRequested`, etc. And then it is parent's responsibility to handle the event. 

Then what happens in `B` is the same. If it can handle the navigation internally (option 2), then the event processing stops there and the navigation is performed using `B`'s `Router`. Otherwise the event continues bubbling up to `A`, where it is finally processed.

By doing the navigation this way you are making each level reusable, because of the proper IoC.

So now let's back to launching `Activities`. Given the structure above it is pretty straightforward how to launch an `Activity`. It will become a 4th option, when `A` can't handle the event and so it bubbles it up to the hosting `Activity` (or maybe another component, again no assumptions here). And the `Activity` is now responsible for the navigation.

Overall I think there should not be too many `Activities` launched (single-Activity should prevail), and so this should be fine. But if this explicitness seems still bloating, then another possible ways could be:

- To handle the navigation in the UI layer. E.g. use the `Context` and launch the `Activity`. Not sure about handling the results, maybe we have something in Jetpack Compose?
- To introduce some sort of `ExternalScreenLauncher` with corresponding methods and ability to subscribe for results. It could be passed via constructors or even added to a custom extended `AppComponentContext`. It could be even published as a plugin to `ComponentContext`, so everyone could use it.

**Reply by moffpage** [reply]
I write single Activity applications too! I was referring to launching external screens like Web Browser, Dialer, Maps, Application Settings, etc.

**Reply by moffpage** [reply]
Damn I forgot that getting a result from activity is so easy with new Activity Result API. In Jetpack Compose, we have ```rememberLauncherForActivityResult(...)``` function, in which we can pass ```ActivityResultContracts.StartActivityForResult()``` as a contract and get an optional ```Intent``` instance by accessing ```data``` property of the result. So yeah, it's pretty simple

**Reply by arkivanov** [reply]
The point here is that components should not be aware of the "kind" of the screen. In the particular example above the component `C` should not care whether the screen being opened on button click is "external" or "internal". There even can be no navigation involved at all, the screen `C` should just not care of it. Its responsibility is to request something from its parent via callback and then process the result.

Like today you are starting an external browser, and tomorrow you will add like a `WebViewComponent`. And the screen is no longer external. With the approach above you will need little to no changes.

But, if you like the idea to handle `Activities` in the UI layer (e.g. via Jetpack Compose as you proposed), then go ahead! Glad there is another solution!

**Reply by moffpage** [reply]
Hi, @arkivanov I've got a question. Why do you need to pass a callback instead of having an activity screen / deep link being a child of the component?

**Reply by arkivanov** [reply]
@moffpage not sure if I understand the question, but an activity can not be a child of a component, because it can not be scoped by the component. Every new activity creates a new root, and navigation between them is managed by the system.

**Reply by moffpage** [reply]
Oh, yeah definitely a good argument right there. I just don't want to start the activity from @Composables (or views, anyway), and I want to share the logic of making a phone call, navigating to a map, etc. Every single system activity start should be handled by root platform-specific objects like Activity and whatever else it is in iOS world?

**Reply by arkivanov** [reply]
But the logic can not be shared between platforms, because it is platform specific. If you really want to start screens directly from components, then you need to provide/inject some API. So create something called e.g. `ExternalScreenLauncher` and provide it to components (via constructor or via extended `AppComponentContext`).

But as I mentioned above, this will make components aware of the "kind" of the screen. Like making a phone call is "external", but showing an item details is "internal". From my point of view components should not care about this sort of stuff.

**Reply by moffpage** [reply]
Good point about components caring about whether their router gonna navigate to an "internal"/"scoped" screen or external (in the way that it's gonna leave the app). However I don't understand your take that logic cannot be shared, couldn't you just say "`router.navigateTo(Dialer(phoneNumber = "..."))`" in your business logic? Oh damn, why can't I decide which one is the right one.

**Reply by arkivanov** [reply]
The `Router` is scoped in its hosting component, it can navigate between components in scope of the host. When you want to navigate to a component in a scope of one of the parents, you have to bubble the event up. Of course you technically can inject parent's `Router` to child components, but this will couple the child component to that exact parent. This will prevent the child and the parent to be placed in separate Gradle modules, and also prevent the child from being reused.

**Reply by moffpage** [reply]
Finally I rethought through this topic and I'm now able to understand your approach. Big thanks to you, Arkadii! What an amazing feedback.
